_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,187 @@
+---
+layout: post
+title: "Make Your Data Obey"
+date: 2016-03-12 10:00:00
+categories: obey data validation modelling
+author: kent.safranski
+cover: /assets/images/covers/obey.jpg
+description: Data is what programming is all about. Applications would be nothing without that beautiful i/o of data moving between modules, services, and data storage. But data is nothing if it's not consistant.
+comments: true
+disqus: true
+---
+
+Data is what programming is all about. Applications would be nothing without that beautiful i/o of data moving between modules, services, and data storage. But data is nothing if it's not consistant. One often ill-attended area of application development is data modelling and validation; most applications have a layer for addressing this, but a lot fall short of really implementing a solid solution.
+
+**[Obey](https://github.com/TechnologyAdvice/obey) was developed specifically to address, and provide solutions for, data modelling and validation for modern web applications developed in JavaScript.**
+
+## Diving In
+
+Let's start with a simple example; validating a user account. There are some core data points we want to ensure, so a model designating these is a good starting point:
+
+{% highlight javascript %}
+import obey from 'obey'
+
+const user = obey.model({
+    id: { type: 'uuid', required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+The above should all be fairly easy to understand. Now let's put it into action, saving to a datasource. Let's pretend this is going into a datasource using a CRUD lib, already setup, called `db`. Using the model above:
+
+{% highlight javascript %}
+import db from './db'
+
+user.validate({
+    id: 'fe6f80c6-c530-4f10-ad0c-89f0f68a879f',
+    email: 'jdoe@email.com',
+    password: 'p@55w0rd'
+})
+// Assumes a db.create method which accepts an object as argument
+.then(db.create).then(res => {
+    console.log('User Created!', res)
+})
+.catch(err => {
+    console.log('Failed!', err)
+})
+{% endhighlight %}
+
+Overall pretty simple stuff. We validate the model, then create the user (or catch the error). But what about that `id`? We probably want to have that generated automatically. Yes, most databases auto-generate unique id's, but play along; this is a perfect example.
```

Comment:
I'd actually disagree with this. Most SQLs, definitely, but for any other type of DB, those that generate their own data are far in the minority.